File: selected_coding_interview/codes/cpp/lc_215_kth_largest_element_in_an_array/lc_215_kth_largest_element_in_an_array_s2.cpp
/*
 * File: lc_215_kth_largest_element_in_an_array_s2.cpp
 * Created Time: 2025-12-30
 * Author: krahets
 */

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return quickSort(nums, k, 0, nums.size() - 1);
    }
private:
    int quickSort(vector<int>& nums, int k, int l, int r) {
        int i = l, j = r;
        while (i < j) {
            while (i < j && nums[j] >= nums[l]) j--;
            while (i < j && nums[i] <= nums[l]) i++;
            swap(nums[i], nums[j]);
        }
        swap(nums[i], nums[l]);
        if (i > nums.size() - k) return quickSort(nums, k, l, i - 1);
        if (i < nums.size() - k) return quickSort(nums, k, i + 1, r);
        // 若基准数索引为 n - k ，则直接返回该元素
        return nums[nums.size() - k];
    }
};

int main() {
    // ======= Test Case =======
    // TODO: Add specific test case

    // ====== Driver Code ======
    Solution* slt = new Solution();
    // TODO: Call solution method and print result

    return 0;
}

File: selected_coding_interview/codes/java/lc_215_kth_largest_element_in_an_array/lc_215_kth_largest_element_in_an_array_s2.java
/*
* File: lc_215_kth_largest_element_in_an_array_s2.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_215_kth_largest_element_in_an_array;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        private void swap(int[] nums, int i, int j) {
            int tmp = nums[i];
            nums[i] = nums[j];
            nums[j] = tmp;
        }
        private int quickSort(int[] nums, int k, int l, int r) {
            int i = l, j = r;
            while (i < j) {
                while (i < j && nums[j] >= nums[l]) j--;
                while (i < j && nums[i] <= nums[l]) i++;
                swap(nums, i, j);
            }
            swap(nums, i, l);
            if (i > nums.length - k) return quickSort(nums, k, l, i - 1);
            if (i < nums.length - k) return quickSort(nums, k, i + 1, r);
            // 若基准数索引为 n - k ，则直接返回该元素
            return nums[nums.length - k];
        }
        public int findKthLargest(int[] nums, int k) {
            return quickSort(nums, k, 0, nums.length - 1);
        }
    }

public class lc_215_kth_largest_element_in_an_array_s2 {
    public static void main(String[] args) {
        // ======= Test Case =======
        int[] test_input_nums = new int[]{3, 2, 1, 5, 6, 4};
        int test_input_k = 2;
        int expected_output = 5;

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.findKthLargest(test_input_nums, test_input_k);
        System.out.println(result);

    }
}
